### PR TITLE
tickets/DM 42420

### PR DESCRIPTION
**Pull request by couger01**
- **Implement tcpip for zaber stage**
- **Add news fragments**

### CODE REVIEW

File: python/lsst/ts/linearstage/controllers/igus_dryve.py [hunk 1]
```diff
@@ -80,25 +83,29 @@ class Igus(Stage):
...
     """
 
-    def __init__(self, config, simulation_mode, log) -> None:
+    def __init__(
+        self, config: types.SimpleNamespace, simulation_mode: int, log: logging.Logger
+    ) -> None:
         super().__init__(config=config, simulation_mode=simulation_mode, log=log)
-        self.commander = None
-        self.mock_ctrl = None  # mock controller, or None of not constructed
+        self.commander: None = None
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/controllers/igus_dryve.py]: It looks like this attribute isn't used anywhere so please remove it.

File: python/lsst/ts/linearstage/controllers/igus_dryve.py [hunk 2]
```diff
@@ -80,25 +83,29 @@ class Igus(Stage):
...
-        self.reader = None
-        self.writer = None
-        self.cmd_lock = asyncio.Lock()
-        self.connection_timeout = 5
-        self.position = None
-        self.status = None
-        self.mode_num = None
+        self.connect_task: asyncio.Task = utils.make_done_future()
+        self.reader: asyncio.StreamReader | None = None
+        self.writer: asyncio.StreamWriter | None = None
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/controllers/igus_dryve.py:97]: Please do not use a reader and writer here but a tcpip Client.

File: python/lsst/ts/linearstage/controllers/igus_dryve.py [hunk 3]
```diff
@@ -156,7 +167,7 @@ async def connect(self):
         # Now verify that controller is enabled (DI7 is high)
         # This must be done from the controller GUI or by sending a 24V
         # signal to the input.
-        response = await self.retrieve_status()
+        response: tuple = await self.retrieve_status()
 
         # Check that byte 20 is has bits 0,1,2 set
         if not (response[20] & 0b111):
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/controllers/igus_dryve.py:174]: I know this line hasn't changed but please extract all magic numbers to constants defined at the top of the file and use them.

File: python/lsst/ts/linearstage/controllers/igus_dryve.py [hunk 4]
```diff
@@ -211,14 +223,14 @@ async def start_mock_ctrl(self):
         )
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/controllers/igus_dryve.py:225]: Not related to this line but please extract the magic number in line 219 to a constant or use the configured timeout value.

File: python/lsst/ts/linearstage/controllers/igus_dryve.py [hunk 5]
```diff
@@ -301,16 +314,16 @@ async def set_drive_settings(self):
...
             raise NotImplementedError(
                 f"Value for given for feed constant of {self.config.feed_rate}"
                 f" exceeds limit of {(2 ** 16) / _multi_factor}"
             )
-        byte19 = _feed_constant & 0b11111111
-        byte20 = _feed_constant >> 8
-        telegram = (
+        byte19: int = _feed_constant & 0b11111111
+        byte20: int = _feed_constant >> 8
+        telegram: tuple = (
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/controllers/igus_dryve.py:330]: These three lines are excellent examples of magic numbers that should be defined in constants.

File: python/lsst/ts/linearstage/controllers/igus_dryve.py [hunk 6]
```diff
@@ -783,10 +798,10 @@ async def move_absolute(self, value):
         await self.send_telegram(telegrams_write["start_motion"])
         # Now poll until target is reached
         await self.poll_until_result(
-            [telegrams_read["target_reached"]], timeout=movement_time + 5
+            [telegrams_read["target_reached"]], timeout=movement_time + 5.0
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/controllers/igus_dryve.py:813]: Again a magic number.

File: python/lsst/ts/linearstage/controllers/zaber_lst.py [hunk 7]
```diff
@@ -19,23 +19,229 @@
...
 
+try:
+    from zaber_motion import Units
+    from zaber_motion.ascii import AxisType, Connection, Device
+    from zaber_motion.exceptions import (
+        CommandFailedException,
+        ConnectionFailedException,
+    )
+except ImportError:
+    pass
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/controllers/zaber_lst.py]: What is supposed to happen if the import fails? Please at least add an inline comment here that explains that and why the except block is empty.

File: python/lsst/ts/linearstage/controllers/zaber_lst.py [hunk 8]
```diff
@@ -19,23 +19,229 @@
...
+
 from .stage import Stage
 
 _LOCAL_HOST = "127.0.0.1"
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/controllers/zaber_lst.py]: Please use the tcpip localhost constant.

File: python/lsst/ts/linearstage/controllers/zaber_lst.py [hunk 9]
```diff
@@ -19,23 +19,229 @@
...
+        super().__init__(config, log, simulation_mode)
+        self.client: Connection | None = None
+        self.mock_server: LinearStageServer | None = None
+        self.position: float | None = None
+        self.device: Device | None = None
+
+    @property
+    def connected(self) -> bool:
+        """Is the client connected?"""
+        # Assume client is connected
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/controllers/zaber_lst.py]: I think this comment would be better placed inside the `if`statement. I assume that the zaber library doesn't have a way to verify that a client is connected so please explicitly state that here too.

File: python/lsst/ts/linearstage/controllers/zaber_lst.py [hunk 10]
```diff
@@ -19,23 +19,229 @@
...
+
+        Parameters
+        ----------
+        value : `float`
+            The amount to move by.
+        """
+        if self.device is None:
+            raise RuntimeError("Device has not been received.")
+        for axis_index in range(self.device.axis_count):
+            axis = self.device.get_axis(axis_index + 1)
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/controllers/zaber_lst.py]: It looks like the axis indices start with 1 and not 0. Please add an inline comment here that explicitly states that or give the correct reason why the `+ 1` is necessary.

Comment (wvreeven) [reply python/lsst/ts/linearstage/controllers/zaber_lst.py]: I see that this same block is repeated several times in the code. Please extract to a function that retrieves the axis and use that so there is no repetitive code.

File: python/lsst/ts/linearstage/controllers/zaber_lst.py [hunk 11]
```diff
@@ -19,23 +19,229 @@
...
+            if axis.axis_type != AxisType.UNKNOWN:
+                try:
+                    await axis.move_relative_async(
+                        position=value, unit=Units.LENGTH_MILLIMETRES
+                    )
+                except CommandFailedException:
+                    self.log.exception("Move relative failed.")
+                    raise
+            else:
+                self.log.info(f"Skipping {axis.axis_number=}.")
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/controllers/zaber_lst.py]: I see that this same if-block is repeated several times in the code, with a different function call. Please extract to a function that retrieves the axis and use that so there is no repetitive code.

File: python/lsst/ts/linearstage/controllers/zaber_lst.py [hunk 12]
```diff
@@ -180,14 +387,14 @@ async def enable_motor(self):
         value : `bool`
             True to enable the motor, False to disable the motor.
         """
-        pass
+        self.log.info("Zaber stage has no brake and so this is a noop.")
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/controllers/zaber_lst.py]: If this class is initialized correctly then the logger is coming from the CSC. That means that this line will be logged into EFD. Is that really what you want? Same for line 397.

File: python/lsst/ts/linearstage/csc.py [hunk 13]
```diff
@@ -99,6 +104,14 @@ async def configure(self, config):
...
         stage_type = instance["stage_type"]
         stage_class = getattr(controllers, stage_type)
+
+        # Go to fault state if dependency is not importable
+        if stage_class == "ZaberV2":
+            if "zaber_motion" not in sys.modules:
+                await self.fault(
+                    code=0,
+                    report="Zaber motion library is not importable, the dependency is not installed.",
+                )
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/csc.py]: This block doesn't belong here but in the `zaber_lst.py` file.

File: python/lsst/ts/linearstage/csc.py [hunk 14]
```diff
@@ -224,7 +244,9 @@ async def handle_summary_state(self):
                     )
                 except RuntimeError as e:
                     err_msg = "Failed to establish connection to component"
-                    await self.fault(code=2, report=f"{err_msg}: {e}")
+                    await self.fault(
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/csc.py:247]: I am not sure if I agree with this way of calling `self.fault`. I will review this after you have made the requested changes.

File: python/lsst/ts/linearstage/csc.py [hunk 15]
```diff
@@ -320,7 +341,10 @@ async def do_moveRelative(self, data):
         self.assert_notmoving("moveRelative")
         self.assert_target_in_range(data.distance, move_type="relative")
         await self.report_detailed_state(LinearStage.DetailedState.MOVINGSTATE)
-        await self.component.move_relative(data.distance)
+        try:
+            await self.component.move_relative(data.distance)
+        except Exception:
+            pass
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/csc.py]: Again an exception is ignored. Please at least add an inline comment explaining why and what the expected behavior is.

File: python/lsst/ts/linearstage/csc.py [hunk 16]
```diff
@@ -332,5 +356,8 @@ async def do_stop(self, data):
             Command data.
         """
         self.assert_enabled("stop")
-        await self.component.stop()
+        try:
+            await self.component.stop()
+        except Exception:
+            pass
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/csc.py]: Again an exception is ignored. Please at least add an inline comment explaining why and what the expected behavior is.

File: python/lsst/ts/linearstage/mocks/mock_zaber_lst.py [hunk 17]
```diff
@@ -19,14 +19,315 @@
...
+        The max size of a word.
+    position : `lsst.ts.simactuators.PointToPointActuator`
+        The encoder position along the stage.
+    axes : `types.SimpleNamespace`
+        The information from each axis.
+    """
+
+    def __init__(self) -> None:
+        self.address: int = 1
+        self.id: int = 30342
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/mocks/mock_zaber_lst.py]: Lots of magic numbers again. Please extract to constants. Also, some of them have very specific values. Is that necessary to make the library work well? If yes, please explain that in inline comments.

File: python/lsst/ts/linearstage/mocks/mock_zaber_lst.py [hunk 18]
```diff
@@ -19,14 +19,315 @@
...
+        Returns
+        -------
+        `str`
+            The reply that was generated.
+        """
+        msg = msg.lstrip("/")
+        msg = msg.split(" ")
+        count = len(msg)
+        request = types.SimpleNamespace()
+        if count == 3:
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/mocks/mock_zaber_lst.py]: Maybe use a match statement instead? Also, please extract the code for each case to a dedicated method to improve readability of the code.

File: python/lsst/ts/linearstage/mocks/mock_zaber_lst.py [hunk 19]
```diff
@@ -19,14 +19,315 @@
...
+        sub_sub_command = kwargs.get("sub_sub_command")
+        field = kwargs["parameters"].split(":")[0]
+        device_address = int(kwargs["device_id"])
+        axis_address = int(kwargs["axis_id"])
+        response = (
+            f"@{device_address:02} {axis_address} {self.message_id:02} RJ IDLE -- "
+        )
+        match sub_command:
+            case "get":
+                match field:
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/mocks/mock_zaber_lst.py]: Please extract to a dedicated method to improve code readability.

File: python/lsst/ts/linearstage/mocks/mock_zaber_lst.py [hunk 20]
```diff
@@ -19,14 +19,315 @@
...
+                    case "zaber.label":
+                        return response + "BADDATA"
+                    case _:
+                        raise NotImplementedError(f"{field} not implemented")
+            case "axis":
+                match sub_sub_command:
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/mocks/mock_zaber_lst.py]: Please extract to a dedicated method to improve code readability.

File: python/lsst/ts/linearstage/mocks/mock_zaber_lst.py [hunk 21]
```diff
@@ -145,88 +474,75 @@ class MockLST:
...
         self.log = logging.getLogger(__name__)
-        self.position = 0
+        self.position = 0.0
         self.status = "IDLE"
         self.device_number = 1
         self.log.info("MockLST created")
 
     def parse_message(self, msg):
-        """Parse and return the result of the message.
+        """Parse the message and return a reply."""
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/mocks/mock_zaber_lst.py]: Please put the Parameters part back.

**Review by wvreeven — CHANGES_REQUESTED** [review]
It is a nice start but there is a lot open for improvement. See my comments.

File: python/lsst/ts/linearstage/enums.py [hunk 22]
```diff
@@ -0,0 +1,26 @@
+__all__ = ["ErrorCode"]
+
+import enum
+
+
+# TODO Move into ts-xml enums.
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/enums.py]: Please refer to a Jira ticket here.

File: python/lsst/ts/linearstage/csc.py [hunk 23]
```diff
@@ -238,10 +248,9 @@ async def handle_summary_state(self):
                 else:
                     self.log.debug("Disabling motor")
                     await self.component.disable_motor()
-            except Exception as e:
+            except Exception:
                 err_msg = "Failed to enable or disable motor"
-                await self.fault(code=2, report=f"{err_msg}: {e}")
-                raise e
+                await self.fault(code=ErrorCode.DISABLE_MOTOR, report=f"{err_msg}")
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/csc.py]: In all other cases you call `self.fault` with both the error message and the error except here. Please do that here too.

File: python/lsst/ts/linearstage/controllers/igus_dryve.py (hunk 1 above)
Comment (wvreeven) [reply python/lsst/ts/linearstage/controllers/igus_dryve.py]: Please remove this attribute or explain why it should remain.

File: python/lsst/ts/linearstage/controllers/igus_dryve.py (hunk 3 above)
Comment (wvreeven) [reply python/lsst/ts/linearstage/controllers/igus_dryve.py:174]: Please add a FIXME like you did for other magic numbers.

File: python/lsst/ts/linearstage/controllers/igus_dryve.py (hunk 4 above)
Comment (wvreeven) [reply python/lsst/ts/linearstage/controllers/igus_dryve.py:225]: Please add a FIXME like you did for other magic numbers.

File: python/lsst/ts/linearstage/controllers/igus_dryve.py (hunk 6 above)
Comment (wvreeven) [reply python/lsst/ts/linearstage/controllers/igus_dryve.py:813]: Please add a FIXME like you did for other magic numbers.

File: python/lsst/ts/linearstage/controllers/zaber_lst.py (hunk 10 above)
Comment (wvreeven) [reply python/lsst/ts/linearstage/controllers/zaber_lst.py]: Please create a new ticket for extracting the remaining repeated block into a new method that takes two arguments (function to execute and error code) as explained on Slack and add TODO/FIXME comments to all methods to which is will apply.

File: python/lsst/ts/linearstage/mocks/mock_zaber_lst.py (hunk 18 above)
Comment (wvreeven) [reply python/lsst/ts/linearstage/mocks/mock_zaber_lst.py]: Thanks for using a match statement here. I still think that extracting the code for each case into a dedicated method will improve readability so please do that.

**Review by wvreeven — CHANGES_REQUESTED** [review]
Lots better! But you skipped several of my comments and I added a few more. Please see to all of them and re-request my review.

File: python/lsst/ts/linearstage/controllers/zaber_lst.py [hunk 24]
```diff
@@ -19,23 +19,244 @@
...
+                await axis.move_absolute_async(
+                    position=value, unit=Units.LENGTH_MILLIMETRES
+                )
+            except CommandFailedException as cfe:
+                self.log.exception("Move absolute rejected.")
+                raise salobj.ExpectedError(f"Move absolute rejected for {cfe.message}")
+
+    async def home(self) -> None:
+        """Home the device, needed to gain awareness of position."""
+        if self.device is None:
```
Comment (wvreeven) [inline python/lsst/ts/linearstage/controllers/zaber_lst.py]: Please add the whole if and for loop as well like this:
```
        if self.device is None:
            raise RuntimeError("Device has not been received.")
        for axis in self.axes:
	        try:
	            await command()
	        except CommandFailedException:
	            self.log.exception(f"{command} rejected.")
	            raise
```

File: python/lsst/ts/linearstage/csc.py (hunk 23 above)
Comment (wvreeven) [reply python/lsst/ts/linearstage/csc.py]: No, I meant to always call self.fault everywhere with `report=f"{err_msg}: {e}` so please do that.

**Review by wvreeven — CHANGES_REQUESTED** [review]
One more time :)

File: python/lsst/ts/linearstage/controllers/zaber_lst.py (hunk 24 above)
Comment (couger01) [reply python/lsst/ts/linearstage/controllers/zaber_lst.py]: I don't think that's going to work with passing the axis method directly.

**Review by wvreeven — APPROVED** [review]
LGTM

**Pull request merged**